### PR TITLE
chore: Upgrade snuba-sdk to 2.0.28

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR upgrades the `snuba-sdk` which also requires an upgrade to `parsimonious`. The update to `parsimonious` required some slight code changes to account for updated internals of the library.

### CODE REVIEW

**Review by joshuarli — APPROVED** [review]
need to address the ci fails in https://github.com/getsentry/pypi/pull/663

**Comment by asottile-sentry** [comment]
> need to address the ci fails in [getsentry/pypi#663](https://github.com/getsentry/pypi/pull/663)

here's your root cause: https://github.com/getsentry/snuba-sdk/commit/b7a64da26d17987fd233351a4d37a883fd7fdd9d#r138544221

**Comment by evanh** [comment]
> here's your root cause: [getsentry/snuba-sdk@b7a64da#r138544221](https://github.com/getsentry/snuba-sdk/commit/b7a64da26d17987fd233351a4d37a883fd7fdd9d#r138544221)

@asottile-sentry @joshuarli Why did this start breaking just for these last few builds? Parsimonious has been in the Snuba SDK for a long time. Seems weird that it would just now start breaking in PyPI.

Also, is there a way the Search team can get notified if these problems happen?

**Comment by asottile-sentry** [comment]
> > here's your root cause: [getsentry/snuba-sdk@b7a64da#r138544221](https://github.com/getsentry/snuba-sdk/commit/b7a64da26d17987fd233351a4d37a883fd7fdd9d#r138544221)
> 
> @asottile-sentry @joshuarli Why did this start breaking just for these last few builds? Parsimonious has been in the Snuba SDK for a long time. Seems weird that it would just now start breaking in PyPI.
> 
> Also, is there a way the Search team can get notified if these problems happen?

it just started breaking because of `__init__.py` shuffles

there isn't really a way to be notified that I know of -- the developer which added parsimonious was notified with a failure but picked the wrong fix breaking the package

**Comment by evanh** [comment]
> there isn't really a way to be notified that I know of -- the developer which added parsimonious was notified with a failure but picked the wrong fix breaking the package

What failure notification was this? Not sure what you mean.

File: src/sentry/api/event_search.py [hunk 1]
```diff
@@ -10,7 +10,6 @@
 
 from django.utils.functional import cached_property
 from parsimonious.exceptions import IncompleteParseError
-from parsimonious.expressions import Optional
```
Comment (iambriccardo) [inline src/sentry/api/event_search.py:13]: `Optional` was removed in `0.10.0` in favor of a function that creates an optional `Quantifier`. For this reason, I had to adapt the logic to consider some text as optional if a `Node` has no matched `text`.

File: requirements-dev-frozen.txt [hunk 2]
```diff
@@ -111,7 +111,7 @@ openapi-spec-validator==0.7.1
 outcome==1.2.0
 packaging==21.3
 parse==1.19.0
-parsimonious==0.9.0
+parsimonious==0.10.0
```
Comment (iambriccardo) [inline requirements-dev-frozen.txt:114]: The new `snuba-sdk` version requires parsimonious to be at least `0.10.0`.

File: src/sentry/snuba/metrics_layer/query.py [hunk 3]
```diff
@@ -343,10 +343,6 @@ def _resolve_timeseries_metadata(
     else:
         mappings[metric.mri] = metric.id
 
-    if not metric.entity:
-        entity = _resolve_metrics_entity(metric.mri)  # This should eventually be done in Snuba
-        metric = metric.set_entity(entity.value)
```
Comment (armenzg) [inline src/sentry/snuba/metrics_layer/query.py:348]: @iambriccardo are the code removals in here intentional?

I see that the code was added by someone else.

Comment (evanh) [reply src/sentry/snuba/metrics_layer/query.py:348]: Yes this is intended. The new version of the SDK doesn't use the `entity`.

File: src/sentry/snuba/metrics_layer/query.py [hunk 4]
```diff
@@ -380,17 +376,6 @@ def fetch_namespace(metric: Metric) -> str:
     return namespaces.pop()
 
 
-def _resolve_metrics_entity(mri: str) -> EntityKey:
-    parsed_mri = parse_mri(mri)
-    if parsed_mri is None:
-        raise InvalidParams(f"'{mri}' is not a valid MRI")
-
-    if parsed_mri.namespace == "sessions":
-        return RELEASE_HEALTH_ENTITIES[parsed_mri.entity]
```
Comment (evanh) [inline src/sentry/snuba/metrics_layer/query.py:389]: Can you also remove the `RELEASE_HEALTH_ENTITIES` and `GENERIC_ENTITIES` dictionaries? They're not needed anymore.

**Review by evanh — APPROVED** [review]
One small comment, otherwise thanks for this!

**Comment by evanh** [comment]
Depends on https://github.com/getsentry/snuba/pull/5525

**Pull request merged**